go/mysqlconn/ldapauthserver/auth_server_ldap.go

```diff
@@ -0,0 +1,148 @@
+		return
+	}
+	ldapAuthServer := newAuthServerLdap()
+
+	data, err := ioutil.ReadFile(*ldapAuthConfigFile)
+	if err != nil {
+		log.Fatalf("Failed to read mysql_ldap_auth_config_file: %v", err)
+	}
+	if err := json.Unmarshal(data, &ldapAuthServer); err != nil {
+		log.Fatalf("Error parsing AuthServerLdap config: %v", err)
+	}
+	mysqlconn.RegisterAuthServerImpl("ldap", ldapAuthServer)
+}
+
+func newAuthServerLdap() *AuthServerLdap {
+	return &AuthServerLdap{}
+}
+
+// UseClearText is always true for AuthServerLdap
+func (asl *AuthServerLdap) UseClearText() bool {
+	return true
+}
+
+// Salt is unimplemented for AuthServerLdap
+func (asl *AuthServerLdap) Salt() ([]byte, error) {
+	panic("unimplemented")
+}
+
+// ValidateHash is unimplemented for AuthServerLdap
+func (asl *AuthServerLdap) ValidateHash(salt []byte, user string, authResponse []byte) (string, error) {
+	panic("unimplemented")
+}
+
+// ValidateClearText connects to the LDAP server over TLS,
+// searches for the user, attempts to bind as that user with the supplied password,
+// and, if so configured, queries for the user's groups, returning them in a
+// comma-separated string.
+// In reality, it runs whatever queries are supplied. See TODO(acharis) for an example.
+// It is recommended that queryUser have read-only privileges on ldapServer
+func (asl *AuthServerLdap) ValidateClearText(username, password string) (string, error) {
+	conn, err := ldap.Dial("tcp", fmt.Sprintf("%s:%d", asl.ldapServer, asl.ldapPort))
+	defer conn.Close()
+	if err != nil {
+		return "", err
+	}
+	// Reconnect with TLS
+	err = conn.StartTLS(&tls.Config{})
+	if err != nil {
+		return "", err
+	}
+
+	// queryUser can be read-only
+	err = conn.Bind(asl.queryUser, asl.queryPassword)
+	if err != nil {
+		return "", err
+	}
+
+	// Search for the given username
+	req := ldap.NewSearchRequest(
+		fmt.Sprintf(asl.queryStr, username),
+		ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false,
+		"(objectClass=organizationalPerson)",
+		[]string{"dn"},
+		nil,
+	)
+
+	res, err := conn.Search(req)
+	if err != nil {
+		return "", err
+	}
+
+	if len(res.Entries) != 1 {
+		return "", errors.New("User does not exist or too many entries returned")
+	}
+
+	userdn := res.Entries[0].DN
+
+	// Bind as the user to verify their password
+	err = conn.Bind(userdn, password)
+	if err != nil {
+		return "", err
+	}
+	if !asl.getGroups {
+		return "", nil
+	}
+
+	// Rebind as the query user for group query
+	err = conn.Bind(asl.queryUser, asl.queryPassword)
+	if err != nil {
+		return "", err
+	}
+
+	req = ldap.NewSearchRequest(
+		asl.groupQueryStr,
+		ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false,
+		fmt.Sprintf("(memberUid=%s)", username),
+		[]string{"cn"},
+		nil,
+	)
+	res, err = conn.Search(req)
+	if err != nil {
+		return "", err
+	}
+	var buffer bytes.Buffer
+	sep := ""
+	for _, entry := range res.Entries {
+		for _, attr := range entry.Attributes {
```

Comment:
@alainjobart thoughts on just adding caching to my approach? It can be size limited and time-based (30s or configurable), or the cache can be periodically refreshed. For instance, if I query in, ldapacl.go will do the ldap query and add me to a map:

```
func (factory *Factory) New(entries []string) (acl.ACL, error) {
  // create LdapAcl
  go lacl.refreshCachedUsers()
}

func (lacl LdapAcl) IsMember(principal string) bool {
  if _, ok := authCache[principal]; ok {
    return ok
  }
  authCache[principal] = lacl.internalIsMember(principal)
  return authCache[principal]
}

func (lacl LdapAcl) internalIsMember(principal string) bool {
 // do ldap queries
}

func (lacl LdapAcl) refreshCachedUsers() {
 // on some timer, iterate authCache and re-run internalIsMember
 // authCache can also have a TTL on it that is refreshed on access
}
```

Or would you still prefer you're approach?